ichnaea/data/station.py

```diff
@@ -553,51 +559,50 @@ def __call__(self):
         if not sharded_obs:
             return
 
-        success = False
-        for i in range(self._retries):
-            try:
-                stats_counter = defaultdict(int)
-                updated_areas = set()
-
-                with self.task.db_session() as session:
-                    for shard, shard_values in sharded_obs.items():
-                        updated_areas.update(
-                            self.update_shard(
-                                session, shard, shard_values, stats_counter
-                            )
-                        )
-
-                success = True
-            except StatementError as exc:
-                if (
-                    isinstance(exc.orig, MySQLError)
-                    and exc.orig.args[0] in self._retriable
-                ):
-                    success = False
-                    METRICS.incr(
-                        "data.station.dberror",
-                        1,
-                        tags=[
-                            "type:%s" % self.station_type,
-                            "errno:%s" % exc.orig.args[0],
-                        ],
-                    )
-                    time.sleep(self._retry_wait * (i ** 2 + 1))
-                else:
-                    raise
-
-            if success:
-                break
+        retry_update_observations = backoff.on_exception(
+            backoff.expo,
+            StatementError,
+            max_tries=3,
+            giveup=self.abort_or_count_exception,
+        )(self.update_observations)
+
+        updated_areas, stats = retry_update_observations(sharded_obs)
+        with self.task.redis_pipeline() as pipe:
+            if updated_areas:
+                self.queue_area_updates(pipe, updated_areas)
+            self.emit_stats(pipe, stats)
+
+        if self.data_queue.ready():
+            self.task.apply_countdown(kwargs={"shard_id": self.shard_id})
+
+    def update_observations(self, sharded_observations):
+        """Update the station data based on per-shard observations."""
+        stats = defaultdict(int)
+        updated_areas = set()
 
-        if success:
-            with self.task.redis_pipeline() as pipe:
-                if updated_areas:
-                    self.queue_area_updates(pipe, updated_areas)
+        with self.task.db_session() as session:
+            for shard, shard_values in sharded_observations.items():
+                areas = self.update_shard(session, shard, shard_values, stats)
+                updated_areas.update(areas)
+        return updated_areas, stats
 
-                self.emit_stats(pipe, stats_counter)
+    def abort_or_count_exception(self, exception):
+        """Check if we should retry or abort on exception.
```

Comment:
I'd rename this to `should_giveup` to continue the language that backoff is using. It's returning a boolean, but the method name suggests it's returning ABORT or COUNT and for me, it requires too much reading to figure out what's going on. I think generating metrics is really a side-effect of the method, so I don't think that needs to be part of the method name.